sql/api/src/main/scala/org/apache/spark/sql/internal/SqlApiConfHelper.scala

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.internal
+
+import java.util.concurrent.atomic.AtomicReference
+
+private[sql] object SqlApiConfHelper {
+  // Shared keys.
+  val ANSI_ENABLED_KEY: String = "spark.sql.ansi.enabled"
+  val LEGACY_TIME_PARSER_POLICY_KEY: String = "spark.sql.legacy.timeParserPolicy"
+  val CASE_SENSITIVE_KEY: String = "spark.sql.caseSensitive"
+  val SESSION_LOCAL_TIMEZONE_KEY: String = "spark.sql.session.timeZone"
+  val LOCAL_RELATION_CACHE_THRESHOLD_KEY: String = "spark.sql.session.localRelationCacheThreshold"
+
+  val confGetter: AtomicReference[() => SqlApiConf] = {
+    new AtomicReference[() => SqlApiConf](() => DefaultSqlApiConf)
+  }
+
+  def getConfGetter: AtomicReference[() => SqlApiConf] = confGetter
```

Comment:
If this line is called in a JVM before any code touching `object SQLConf`, are we going to return `DefaultSqlApiConf` which is wrong?

Reply:
@hvanhovell Do you know if the scenario mentioned here could happen?

Reply:
@zsxwing good point! How about this: We make `SqlApiConf` the only place to deal with the active conf problem. It should use reflection to get the active `SQLConf` and set it as the getter. `SQLConf` should only access this new `SqlApiConfHelper` for config names.

Reply:
@zsxwing I don't think it is an issue. An end user should never touch this object unless they know what they are doing. That is a limitation of this approach, we should document this, and we can add a liniting rule if we must.

Just to make sure that we are on the same page. In a normal situation, the initialisation is as follows:

- User touches the `SQLConf` object. This loads and initialises the `SQLConf` object. The loading of the `SQLConf` object triggers the initialisation of the `SqlApiConfHelper` because it needs to set the setter, and it needs to load the shared keys.
- User touces the `SqlApiConf` object. This tries to load the `SQLConf` object as part of its initialisation (see previous bullet), if it is on the classpath the getter will be updated to return the `SQLConf`.